### PR TITLE
[two_dimensional_scrollables] Fixes TreeViewNode collapsing not working

### PR DESCRIPTION
**Pull request by Mairramer**
Fixes [154295](https://github.com/flutter/flutter/issues/154295)
This change addresses bug [#154295](https://github.com/flutter/flutter/issues/154295) by ensuring that nodes are handled correctly when closing their children.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[Discord]: https://github.com/flutter/flutter/blob/master/docs/contributing/Chat.md
[linked to at least one issue that this PR fixes]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#overview
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#version
[following repository CHANGELOG style]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changelog-style
[exempt from CHANGELOG changes]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changelog
[test-exempt]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#tests

### CODE REVIEW

File: packages/two_dimensional_scrollables/test/tree_view/tree_test.dart [hunk 1]
```diff
@@ -718,6 +718,101 @@ void main() {
       expect(treeView.treeNodeBuilder, isA<TreeViewNodeBuilder<String>>());
       expect(treeView.treeRowBuilder, isA<TreeViewRowBuilder<String>>());
     });
+
+    testWidgets(
+        'TreeSliverNode should close all children when collapsed when animation is completed',
```
Comment (Piinks) [inline packages/two_dimensional_scrollables/test/tree_view/tree_test.dart:723]: Can you update the language here? This seems to be referring to the 1D version in the framework.

**Comment by Piinks** [comment]
Can you also update the title of this PR to reflect this package? It currently refers to the framework change, not this one. Also, the PR template for this repo is different from the framework. For example, it specifies the title of the PR should start with the [name_of_package].

File: packages/two_dimensional_scrollables/lib/src/tree_view/tree.dart [hunk 2]
```diff
@@ -940,9 +948,7 @@ class _TreeViewState<T> extends State<TreeView<T>>
           controller.forward();
         case false:
           // Collapsing
-          controller.reverse().then((_) {
-            _unpackActiveNodes();
-          });
+          controller.reverse();
```
Comment (lbgitjp) [inline packages/two_dimensional_scrollables/lib/src/tree_view/tree.dart:951]: hwo about change below code
https://github.com/flutter/packages/blob/bb53e5d536aeba377b5f8877ad735973293ae200/packages/two_dimensional_scrollables/lib/src/tree_view/tree.dart#L936-L945
to

```
      switch (node._expanded) {
        case true:
          // Expanding
          controller.forward();
        case false:
          // Collapsing
          controller.reverse();
      }
```

File: packages/two_dimensional_scrollables/lib/src/tree_view/tree.dart [hunk 3]
```diff
@@ -902,6 +902,14 @@ class _TreeViewState<T> extends State<TreeView<T>>
...
               _currentAnimationForParent.remove(node);
               _updateActiveAnimations();
+              // If the node is collapsing, we need to rebuild the active nodes
+              // to reflect the new state.
+              // This is because the node's children are no longer active.
+              if (!node._expanded) {
+                setState(() {
+                  _unpackActiveNodes();
+                });
+              }
```
Comment (lbgitjp) [inline packages/two_dimensional_scrollables/lib/src/tree_view/tree.dart:912]: how about remove 'if' and chang the code to below:
```
              setState(() {
                _unpackActiveNodes();
              });
```

Comment (lbgitjp) [inline packages/two_dimensional_scrollables/lib/src/tree_view/tree.dart:912]: how about remove 'if' and chang the code to below:
```
              setState(() {
                _unpackActiveNodes();
              });
```

File: packages/two_dimensional_scrollables/test/tree_view/tree_test.dart [hunk 4]
```diff
@@ -718,6 +718,101 @@ void main() {
       expect(treeView.treeNodeBuilder, isA<TreeViewNodeBuilder<String>>());
       expect(treeView.treeRowBuilder, isA<TreeViewRowBuilder<String>>());
     });
+
+    testWidgets(
```
Comment (Piinks) [inline packages/two_dimensional_scrollables/test/tree_view/tree_test.dart:722]: Similar to the framework change, this test does not fail when run without the code change. It should so it verifies the fix works as expected, and prevents us from regressing in the future.

**Review by Piinks — CHANGES_REQUESTED** [review]
This will need updated tests to move forward.

**Comment by Mairramer** [comment]
@Piinks Same of comment https://github.com/flutter/flutter/pull/153438#issuecomment-2363614393.

File: packages/two_dimensional_scrollables/test/tree_view/tree_test.dart [hunk 5]
```diff
@@ -814,6 +814,74 @@ void main() {
...
+        ),
+      ));
+
+      expect(find.text('alpha'), findsOneWidget);
+      expect(find.text('uno'), findsOneWidget);
+      expect(find.text('dos'), findsOneWidget);
+      expect(find.text('tres'), findsOneWidget);
+
+      // Using runAsync to handle collapse and animations properly.
+      await tester.runAsync(() async {
```
Comment (Piinks) [inline packages/two_dimensional_scrollables/test/tree_view/tree_test.dart:876]: This change made this a valid regression test, can you share why it needed to runAsync?

Comment (Piinks) [reply packages/two_dimensional_scrollables/test/tree_view/tree_test.dart:876]: Answered in related PR https://github.com/flutter/flutter/pull/153438#discussion_r1801888873

**Review by Piinks — APPROVED** [review]
LGTM

**Pull request merged**